imageprocessing/src/main/java/com/example/imageprocessing/PreProcessing.java

```diff
@@ -0,0 +1,394 @@
+
+                    //Verify if the image is good enough
+                    if(isBright(modifiedMat) == BrightnessValue.IMAGE_IS_OK){
+                        //If the conversion failed it returns the original image
+                        try{
+                            return IPUtils.conversionMatToBitmap(modifiedMat);
+                        } catch (ConversionFailedException error){
+                            Log.e(TAG, error.getErrorMessage());
+                            return image;
+                        }
+                    }
+                }
+                break;
+
+            case IMAGE_IS_OK: //Image is neither too bright nor too dark
+                Log.d(TAG, "Case==IMAGE_IS_OK");
+                return image;
+        }
+        return image;
+    }
+
+
+    /**
+     * @author Thomas Porro(g1), Oscar Garrido (g1), Giovanni Fasan(g1).
+     * See PreProcessingMethods.java
+     */
+    @Override
+    public boolean isBlurred(Bitmap image) {
+
+        //Total number of color
+        int maxLap = -16777216;
+
+        //Threshold above which the color is out of focus
+        final int threshold = -6118750;
+
+        //Converts the image into a matrix
+        Mat imageMat;
+        try {
+            imageMat = IPUtils.conversionBitmapToMat(image);
+        } catch (ConversionFailedException error){
+            Log.e(TAG, error.getErrorMessage());
+            return false;
+        }
+
+        //Turn the colored matrix into a grayscale matrix
+        Mat grayImageMat = new Mat();
+        Imgproc.cvtColor(imageMat, grayImageMat, Imgproc.COLOR_BGR2GRAY);
+
+        /*Use the openCV's Laplacian methods to apply a transformation that allow us to detect
+          the image blurriness*/
+        Mat laplacianMat = new Mat();
+        Imgproc.Laplacian(grayImageMat, laplacianMat, CV_8U);
+        //Converts the matrix into another format used to detect the blur
+        Mat laplacianMat8Bit = new Mat();
+        laplacianMat.convertTo(laplacianMat8Bit, CV_8UC1);
+
+        //Create a Bitmap with the given matrix, and obtain all the pixels from it
+        Bitmap laplacianImage;
+        try{
+            laplacianImage = IPUtils.conversionMatToBitmap(laplacianMat8Bit);
+        } catch (ConversionFailedException error){
+            Log.e(TAG, error.getErrorMessage());
+            return false;
+        }
+
+        //Extracts all the pixels of the laplacian image into the array
+        int[] pixels = IPBuilder.doGetPixels(new IPBuilder.GetPixelsBuilder(laplacianImage)
+                  .withStride(laplacianImage.getWidth())
+                  .withWidth(laplacianImage.getWidth())
+                  .withHeight(laplacianImage.getHeight())
+                  );
+                  
+        //Searches the pixel that has the highest colour range in the RGB format
+        for(int pixel : pixels){
+            if(pixel > maxLap){
+                maxLap = pixel;
+            }
+        }
+
```

Comment:
If I'm not wrong you determine if an image is or isn't blurred looking to a single pixel value. Is this correct? Don't you have to look to a larger group of pixels to make this decision?

Reply:
We read in various articles that a way to estimate the sharpness of an image is to apply a Laplace filter and then pick the maximum value. I'll add this information as a comment in the code